File: camera-core/src/akvc/sdk/virtual_camera.py
# SPDX-License-Identifier: Apache-2.0
"""High-level virtual camera wrapper for external apps."""

from __future__ import annotations

from pathlib import Path
from typing import Optional

import numpy as np

from akvc.core.frame import Frame
from akvc.core.frame_pipeline import (
    ColorConvertStage,
    FpsRegulator,
    FramePipeline,
    ResizeStage,
)
from akvc.core.frame_sink import FrameSink, create_sink
from akvc.core.helper.client import HelperService


class VirtualCamera:
    """Push BGR frames into the system virtual camera."""

    def __init__(
        self,
        *,
        width: int = 1280,
        height: int = 720,
        fps: float = 30.0,
        helper_exe: str | Path | None = None,
        pipeline: FramePipeline | None = None,
    ) -> None:
        self.width = width
        self.height = height
        self.fps = fps
        self._helper = HelperService(helper_exe=helper_exe)
        self._pipeline = pipeline or (
            FramePipeline()
            .add(ResizeStage(target_w=width, target_h=height))
            .add(FpsRegulator(target_fps=fps))
            .add(ColorConvertStage(dst="NV12"))
        )
        self._sink: Optional[FrameSink] = None
        self._started = False
        self._mf_registered = False

    @property
    def started(self) -> bool:
        return self._started

    @property
    def consumer_count(self) -> int:
        if self._sink is None:
            return 0
        return self._sink.consumer_count

    def start(self, name: str = "AK Virtual Camera") -> None:
        if self._started:
            return
        if not self._helper.start():
            raise RuntimeError("failed to start akvc helper")
        if not self._helper.ping():
            raise RuntimeError("akvc helper is not responding")
        if not self._mf_registered:
            if not self._helper.register_mf(name=name):
                raise RuntimeError("failed to register MF virtual camera")
            self._mf_registered = True
        sink = create_sink()
        sink.open()
        self._sink = sink
        self._started = True

    def push_frame(self, bgr: np.ndarray) -> None:
        if not self._started or self._sink is None:
            raise RuntimeError("virtual camera is not started")
        frame = Frame.from_bgr(bgr)
        frame = self._pipeline.process(frame)
        self._sink.publish(frame)

    def stop(self) -> None:
        if not self._started:
            return
        assert self._sink is not None
        self._sink.close()
        self._sink = None
        self._started = False

    def close(self) -> None:
        self.stop()
        self._helper.stop()

    def shutdown(self) -> None:
        self.close()

    def __enter__(self) -> "VirtualCamera":
        self.start()
        return self

    def __exit__(self, exc_type, exc, tb) -> None:
        self.close()

File: camera-core/src/virtual_camera.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 AK Virtual Camera Authors
//
// VirtualCamera facade — platform dispatch via pimpl.

#include "akvc/virtual_camera.h"

#if defined(_WIN32)
#  include "akvc/platform/windows/windows_session.h"
#elif defined(__APPLE__)
#  include "akvc/platform/macos/macos_session.h"
#else
#  error "akvc::VirtualCamera: unsupported platform"
#endif

namespace akvc {

#if defined(_WIN32)
struct VirtualCamera::Impl {
    windows::WindowsVirtualCameraSession session;
    Impl(int w, int h, double fps, const std::string& camera_name, const std::string& helper_exe)
        : session(w, h, fps, helper_exe, camera_name) {}
};
#elif defined(__APPLE__)
struct VirtualCamera::Impl {
    macos::MacVirtualCameraSession session;
    Impl(int w, int h, double fps, const std::string& camera_name, const std::string& helper_exe)
        : session(w, h, fps, camera_name) { (void)helper_exe; }
};
#endif

VirtualCamera::VirtualCamera(int width, int height, double fps,
                             std::string camera_name, std::string helper_exe)
    : impl_(std::make_unique<Impl>(width, height, fps, camera_name, helper_exe)) {}

VirtualCamera::~VirtualCamera() = default;

Status VirtualCamera::start() { return impl_->session.start(); }

Status VirtualCamera::push_frame(const FrameInput& frame) {
    return impl_->session.push_frame(frame);
}

void VirtualCamera::stop() { impl_->session.stop(); }

bool VirtualCamera::started() const { return impl_->session.started(); }

int VirtualCamera::consumer_count() const { return impl_->session.consumer_count(); }

const char* VirtualCamera::last_error() const { return impl_->session.last_error(); }

}  // namespace akvc
